src/Microsoft.Diagnostics.Monitoring/DiagnosticsEventPipeProcessor.cs

```diff
@@ -0,0 +1,320 @@
+
+using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Diagnostics.Tracing;
+using Microsoft.Diagnostics.Tracing.Parsers.FrameworkEventSource;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring
+{
+    public enum PipeMode
+    {
+        Logs = 1,
+        Metrics,
+    }
+
+    public class DiagnosticsEventPipeProcessor : IAsyncDisposable
+    {
+        private readonly ILoggerFactory _loggerFactory;
+        private readonly IEnumerable<IMetricsLogger> _metricLoggers;
+        private readonly PipeMode _mode;
+
+        //These values don't change so we compute them only once
+        private readonly List<string> _dimValues;
+
+        public const string NamespaceName = "Namespace";
+        public const string NodeName = "Node";
+        private static readonly List<string> DimNames = new List<string> { NamespaceName, NodeName };
+
+        public DiagnosticsEventPipeProcessor(ContextConfiguration contextConfig, PipeMode mode,
+            ILoggerFactory loggerFactory,
+            IEnumerable<IMetricsLogger> metricLoggers)
+        {
+            _dimValues = new List<string> { contextConfig.Namespace, contextConfig.Node };
+            _metricLoggers = metricLoggers;
+            _mode = mode;
+            _loggerFactory = loggerFactory;
+        }
+
+        public async Task Process(int pid, int duration, CancellationToken token)
+        {
+            await Task.Factory.StartNew(async () =>
+            {
+                EventPipeEventSource source = null;
+                DiagnosticsMonitor monitor = null;
+                try
+                {
+                    MonitoringSourceConfiguration config = null;
+                    if (_mode == PipeMode.Logs)
+                    {
+                        config = new LoggingSourceConfiguration();
+                    }
+                    if (_mode == PipeMode.Metrics)
+                    {
+                        config = new MetricSourceConfiguration();
+                    }
+
+                    monitor = new DiagnosticsMonitor(config);
+                    Stream sessionStream = await monitor.ProcessEvents(pid, duration, token);
+                    source = new EventPipeEventSource(sessionStream);
+                    
+                    if (_mode == PipeMode.Metrics)
+                    {
+                        // Metrics
+                        HandleEventCounters(source);
+                    }
+
+                    if (_mode == PipeMode.Logs)
+                    {
+                        // Logging
+                        HandleLoggingEvents(source);
+                    }
+
+                    source.Process();
```

Comment:
Whenever I see this blocking code I'll ping @noahfalk. Can we at least file an issue to look at fixing it at some point?

Reply:
We can file an issue (https://github.com/microsoft/perfview/issues/1164), but its probably a large enough amount of work that we wouldn't do it without first having some confidence via a perf measurement that it was likely to make a significant difference. Certainly it won't scale well if you had to do 100s or 1000s of parallel traces, but in practice when this API is used parallelism is usually 1.